### DISCUSSION TITLE
Can't list files in storage ( and other questions ) ?

### DISCUSSION THREAD

**Discussion by ethanchiasson** [supabase/supabase#2411] [Questions] [answered]
I am currently able to upload files to storage. I am trying to list all files within my storage, but am unable to get a response other than `undefined`.

` async function getFiles() {
    try {
      const { data, error } = await supabase.storage.from("files").list();
    } catch (error) {
      alert(error.message);
    } finally {
      loading = false;
    }
  }`
  
  and then in my UI  ( using svelte )...
  
  `{#await getFiles() }
  ...Loading Files
  {:then response}
  {JSON.stringify(response)}
  {/await}`
  
  The following code sample will return undefined. I have a few questions...
  
  1. Do I have to "download" files before I can list them ? If so, this might be the reason the function returns undefined.
  
  2. I am allowing the user to upload multiple files for a specific user in our database ( think Purchase Order PDFs / Images ). That being said, is there a way to `list()` all files that contain X, where X is a string. The reason I am asking is, all files will be named `[User's ID]-[insert random number here]` ( so that I am able to store multiple files for a single user without naming issues ), and thus I would want to return specific files for a specific user by doing something like `.list().contains([User's ID]).
  
Thank you in advance for the help / advice.

**Comment by dshukertjr** [comment, marked as answer]
Hi @ethanchiasson!

Could you try this code without wrapping it in `try catch`? You should see an error being logged if the data is `null`. My guess is that the user does not have rights to access the storage objects because of row level security, but the error log should tell us a bit more about what's happening. 
```typescript
  const { data, error } = await supabase.storage.from('files').list()
  console.log('data', data)
  console.error('error', error)
```

If it was a row level security issue, you can grant access to the user by creating a new security policy. You can find some example policies [here](https://supabase.io/docs/guides/storage#policy-examples). 

Here are answers to your questions:

> Do I have to "download" files before I can list them ? If so, this might be the reason the function returns undefined.

No, you do not have to download the files before you can list them. You should be able to 

>  am allowing the user to upload multiple files for a specific user in our database ( think Purchase Order PDFs / Images ). That being said, is there a way to list() all files that contain X, where X is a string. The reason I am asking is, all files will be named [User's ID]-[insert random number here] ( so that I am able to store multiple files for a single user without naming issues ), and thus I would want to return specific files for a specific user by doing something like `.list().contains([User's ID]).

Instead of doing something like `.list().contains([user's ID])`, you can pass the path of the files to `.list()` method. You might have to do some restructuring of your folders, but let's say every user stored their files under `[User's ID]/filename`. Then you could do this to get the user's files:
```typescript
  const { data, error } = await supabase.storage.from('files').list('[User's ID]')
```

You can read more about what options are available on `.list()` [here](https://supabase.io/docs/reference/javascript/storage-from-list)! 

Hopefully this will clarify the issue you are facing, but if not, please don't hesitate to let me know!

**Reply by ethanchiasson** [reply]
Okay so I attempted to remove the try / catch and I get `null` on the error, and also an array object `data` containing 2 objects ( one of them being my untitled folder ).

Oddly, I am trying to access it like...

`{#await getFiles()}...loading{:then response}
  {#each response.data as file}
    {file.name}
  {/each}
  {/await}`
  
 ...but now instead of `undefined`, my I get `...loading` in my UI. Maybe I am missing something here.
 
 
Edit: So I simply added `return data;` to my `getFiles()` function and I am now able to read the files in the DB. I was attempting to read off the response, as that I how I am doing it with my other ( non-storage ) supabase functions.

Thank you for the help!

**Reply by dshukertjr** [reply]
So glad to hear that you were able to get it to work 🚀🚀🚀

**Reply by ethanchiasson** [reply]
I do have another question...

I am able to upload files into new folders for each user, however, I am a little lost on downloading..

Here I am trying to pass in the `power_number` ( folder name ) and then the file's name + extension `ex.png` by passing in the file object...

```js
async function getFileImage(fileToGet) {
   const { data, error } = await supabase
  .storage
  .from('files')
  .download(`/${params.power_number}/${fileToGet.name}`).then(res => {
    console.log("get File res data =>", res.data)
  })
    console.error('error', error)
  }
  ```
  
I am getting `null` as the response. I am wondering, am I correctly passing in the path ? I tried passing the path into the `.from()` and `.download` and can't seem to get a response.


Edit:

I found part of my error. I was adding `/` before the `${power_number}`, which is not needed, resulting in the path `//..`.

The following function works...

```js
  async function getFileImage(fileToGet) {
   const { data, error } = await supabase
  .storage
  .from('files')
  .download(`${params.power_number}/${fileToGet.name}`).then(res => {
    console.log("get File res data =>", res)
  })
    console.error('error', error)
  }
```

which results in a object with a blob. From here I will attempt to download the image and review it in the UI, while also adding a way to download the file itself.

**Reply by ethanchiasson** [reply]
I apologize for my formatting. I will try to edit my question with better formatting.

**Reply by dshukertjr** [reply]
Hi @ethanchiasson!

Although this example is with next.js and not with Svelte, it might help as it is doing something very similar! https://github.com/supabase/supabase-js/blob/master/example/next-storage/components/Avatar.tsx#L14

Also, when you are using `async/await`, you don't have to use `then` method to get the data! You can just get the data like this:
```typescript
async function getFileImage(fileToGet) {
   const { data, error } = await supabase
    .storage
    .from('files')
    .download(`${params.power_number}/${fileToGet.name}`)
    console.log("get File data =>", data)
    console.error('error', error)
  }
```

Hopefully this helps!
